### PR TITLE
Function manager

### PR DESCRIPTION
**Pull request by yuxtang-amazon**
## Relevant Issues
- [Closes] Issue #1120 

## Description
This request aims to develop a FunctionManager for function overloading and redefine ExprFunction. FunctionManager can house all ExprFunctions as a list, with a 'get' function to check name, arity, and types, and is implemented in evalCompilers and PhysicalPlanCompilerImpl for function compilation.
ExprFunction only retains the CallWithRequired function call and omit variadicParameters and optionalParameters from the functionSignature.
Lastly, I have updated built-in functions(including functions in partiql-lang and partiql-cli) to utilize the new overloading feature and discarded optional/variadic parameters usage.

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
YES

- Any backward-incompatible changes? **[YES/NO]**
YES
Since we've transitioned from functionMap to FunctionManager and redefined ExprFunctions, I updated our built-functions by removing optional/variadic parameters and adding overloaded functions with identical names but varying types/numbers of parameters.
For compilation, I replaced the function map with a function list and use FunctionManager to consolidate the functions.
For StaticTypeInferencer, I used a for loop to loop all candidate functions and thus return the return type.
Finally, I adjusted test cases with variadic parameters, converting them from variadic parameters to a list value of such parameters.

- Any new external dependencies? **[YES/NO]**
NO

- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **[YES/NO]**
YES

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-cli/src/main/kotlin/org/partiql/cli/functions/QueryDDB.kt [hunk 1]
```diff
@@ -47,7 +47,6 @@ class QueryDDB(private val ion: IonSystem) : ExprFunction {
     override val signature = FunctionSignature(
         name = "query_ddb",
         requiredParameters = listOf(StaticType.STRING),
-        optionalParameter = StaticType.NULL_OR_MISSING,
```
Comment (johnedquinn) [inline partiql-cli/src/main/kotlin/org/partiql/cli/functions/QueryDDB.kt:50]: Nice catch here.

File: partiql-cli/src/main/kotlin/org/partiql/cli/functions/ReadFile.kt
Comment (johnedquinn) [inline partiql-cli/src/main/kotlin/org/partiql/cli/functions/ReadFile.kt:1]: Can we make this a bit cleaner? Maybe extract the core functionality in an abstract class:
```kotlin
internal abstract class ReadFile : ExprFunction {
    private fun readHandlers = ...

    internal class ReadFile_1 : ReadFile() {
        ...
    }

    internal class ReadFile_2: ReadFile() {
        ...
    }
}
```

Also, we've lost the ability to use other read handlers besides `ion` when we don't pass in the optional struct. Seemingly, the CSV read handler allows for default values when the bindings object is empty.

File: partiql-cli/src/main/kotlin/org/partiql/cli/functions/WriteFile.kt
Comment (johnedquinn) [inline partiql-cli/src/main/kotlin/org/partiql/cli/functions/WriteFile.kt:1]: Same comment as `ReadFile.kt`.

File: partiql-cli/src/test/kotlin/org/partiql/cli/functions/ReadFileTest.kt [hunk 2]
```diff
@@ -84,7 +85,8 @@ class ReadFileTest {
         val path = getResourcePath("data.ion")
         val args = listOf("\"$path\"").map { it.exprValue() }
         assertThrows<IllegalStateException> {
-            function.callWithOptional(session, args, "{type:\"ion\"}".exprValue())
+            function.callWithRequired(session, args)
```
Comment (johnedquinn) [inline partiql-cli/src/test/kotlin/org/partiql/cli/functions/ReadFileTest.kt:88]: I'm not sure which of these throws the `IllegalStateException`. If they both do, can you make 2 separate tests? Or assert these sequentially?

File: partiql-lang/src/main/kotlin/org/partiql/lang/CompilerPipeline.kt [hunk 3]
```diff
@@ -163,6 +163,8 @@ interface CompilerPipeline {
          */
         fun addFunction(function: ExprFunction): Builder = this.apply { customFunctions[function.signature.name] = function }
 
+//        private fun functions(fm: FunctionManager): Builder = this.apply { functionManager = fm }
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/CompilerPipeline.kt:166]: Remove

File: partiql-lang/src/main/kotlin/org/partiql/lang/CompilerPipeline.kt [hunk 4]
```diff
@@ -163,6 +163,8 @@ interface CompilerPipeline {
          */
         fun addFunction(function: ExprFunction): Builder = this.apply { customFunctions[function.signature.name] = function }
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/CompilerPipeline.kt:164]: Seemingly, we need to modify the definition of `customFunctions` to not be a map, but to also be a list. We're accidentally overwriting custom functions when they share the same name.

File: partiql-lang/src/main/kotlin/org/partiql/lang/ast/passes/inference/StaticTypeInferencer.kt [hunk 5]
```diff
@@ -19,13 +19,13 @@ import org.partiql.types.StaticType
...
  * perspective.
- * @param customFunctionSignatures Custom user-defined function signatures that can be called by the query.
+ * @param customFunction Custom user-defined functions that can be called by the query.
  * @param customTypedOpParameters Mapping of custom type name to [TypedOpParameter] to be used for typed operators
  * (i.e CAST/IS) inference.
  */
 class StaticTypeInferencer(
     private val globalBindings: Bindings<StaticType>,
-    private val customFunctionSignatures: List<FunctionSignature>,
+    private val customFunction: List<ExprFunction>,
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/ast/passes/inference/StaticTypeInferencer.kt:28]:
```suggestion
    private val customFunctions: List<ExprFunction>,
```

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/ExprFunction.kt [hunk 6]
```diff
@@ -47,42 +42,8 @@ interface ExprFunction {
...
-     * [EvaluatingCompiler] validates the arguments before calling [ExprFunction]s.
-     * Hence the implementations are not required to deal with it.
-     *
-     * @param session The current [EvaluationSession].
-     * @param required The required arguments.
-     * @param variadic The variadic arguments.
-     */
-    fun callWithVariadic(session: EvaluationSession, required: List<ExprValue>, variadic: List<ExprValue>): ExprValue {
-        // Deriving ExprFunctions must implement this if they have a valid call form including required parameters and variadic
+    fun callWithRequired(session: EvaluationSession, required: List<ExprValue>): ExprValue {
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/ExprFunction.kt:46]: There's an extra newline above this.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/ExprFunction.kt [hunk 7]
```diff
@@ -98,9 +59,4 @@ interface ExprFunction {
  * @param session The current [EvaluationSession].
  * @param args The argument list.
  */
-fun ExprFunction.call(session: EvaluationSession, args: Arguments): ExprValue =
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/ExprFunction.kt:101]: Please remember to label this as a breaking change in the CHANGELOG.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/ExprFunction.kt [hunk 8]
```diff
@@ -17,11 +17,6 @@ package org.partiql.lang.eval
 import org.partiql.lang.errors.ErrorCode
 import org.partiql.lang.types.FunctionSignature
 
-sealed class Arguments
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/ExprFunction.kt:20]: CHANGELOG breaking change.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/DynamicLookupExprFunction.kt [hunk 9]
```diff
@@ -105,3 +102,13 @@ class DynamicLookupExprFunction : ExprFunction {
...
 }
+
+class DynamicLookupExprFunction2 : ExprFunction {
+
+    override val signature = FunctionSignature(
+        name = DYNAMIC_LOOKUP_FUNCTION_NAME,
+        // Required parameters are: variable name, case sensitivity and lookup strategy
+        requiredParameters = listOf(StaticType.SYMBOL, StaticType.SYMBOL, StaticType.SYMBOL),
+        returnType = StaticType.ANY
+    )
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/DynamicLookupExprFunction.kt:113]: Seemingly, we need to still look in the session's globals:
```kotlin
val found = when {
            globalsFirst -> {
                session.globals[bindingName] ?: searchLocals(variadic, bindingName)
            }
            else -> {
                searchLocals(variadic, bindingName) ?: session.globals[bindingName]
            }
        }

        if (found == null) {
            // We don't know the metas inside ExprFunction implementations.  The ThunkFactory error handlers
            // should add line & col info to the exception & rethrow anyway.
            throwUndefinedVariableException(bindingName, metas = null)
        } else {
            return found
        }
```

And we don't need to search through the locals (`searchLocals`). But, if we can't find in the globals, we still need to throw that same error.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/DynamicLookupExprFunction.kt
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/DynamicLookupExprFunction.kt:1]: Please add these breaking changes to the CHANGELOG.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/ScalarBuiltinsExt.kt
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/ScalarBuiltinsExt.kt:1]: Shared logic via abstract class?

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/ScalarBuiltinsSql.kt [hunk 10]
```diff
@@ -350,6 +359,70 @@ internal object ExprFunctionUpper : ExprFunction {
     }
 }
 
+/**
+ * This function is to test overloading.
+ */
+internal object ExprFunctionPower : ExprFunction {
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/builtins/ScalarBuiltinsSql.kt:365]: While this was good for testing, we don't want to introduce any of the `"query_power"` functions to the Scalar Builtins. We can delete these objects.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+                errorList.add(e)
+            }
+        }
+        throw errorList.first()
+    }
+
+    /**
+     * Check argument types by requiredArgs, OptionalArgs and VariadicArgs
+     */
+//    internal fun checkArgumentTypes(signature: FunctionSignature, args: List<ExprValue>) {
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt:51]: Remove

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt [hunk 12]
```diff
@@ -0,0 +1,115 @@
+package org.partiql.lang.eval.impl
+
+import org.partiql.lang.eval.ExprFunction
+import org.partiql.lang.eval.ExprValue
+import org.partiql.lang.eval.errInvalidArgumentType
+import org.partiql.lang.types.FunctionSignature
+import org.partiql.lang.types.StaticTypeUtils
+import org.partiql.types.StaticType
+
+data class ResultFormat(val function: ExprFunction?, val errorMessage: String?, val arity: Pair<Int, Int>)
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt:10]: Can we make this a child class of FunctionManager?

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt [hunk 13]
```diff
@@ -76,7 +78,8 @@ import org.partiql.types.SymbolType
  */
 internal class StaticTypeInferenceVisitorTransform(
     globalBindings: Bindings<StaticType>,
-    customFunctionSignatures: List<FunctionSignature>,
+//    customFunctionSignatures: List<FunctionSignature>,
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt:81]: Remove

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt [hunk 14]
```diff
@@ -94,8 +97,9 @@ internal class StaticTypeInferenceVisitorTransform(
     }
 
     /** The built-in functions + the custom functions. */
-    private val allFunctions: Map<String, FunctionSignature> =
-        SCALAR_BUILTINS_DEFAULT.associate { it.signature.name to it.signature } + customFunctionSignatures.associateBy { it.name }
+//    private val allFunctions: Map<String, FunctionSignature> =
+//        SCALAR_BUILTINS_DEFAULT.associate { it.signature.name to it.signature } + customFunctionSignatures.associateBy { it.name }
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt:101]: Remove

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt [hunk 15]
```diff
@@ -76,7 +78,8 @@ import org.partiql.types.SymbolType
  */
 internal class StaticTypeInferenceVisitorTransform(
     globalBindings: Bindings<StaticType>,
-    customFunctionSignatures: List<FunctionSignature>,
+//    customFunctionSignatures: List<FunctionSignature>,
+    val customFunction: List<ExprFunction>,
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt:82]:
```suggestion
    val customFunctions: List<ExprFunction>,
```

File: partiql-lang/src/test/kotlin/org/partiql/lang/eval/QueryPowerTest.kt
Comment (johnedquinn) [inline partiql-lang/src/test/kotlin/org/partiql/lang/eval/QueryPowerTest.kt:1]: Remove these since `query_power` is being removed.

File: partiql-lang/src/test/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransformTest.kt [hunk 16]
```diff
@@ -7191,8 +7190,9 @@ class StaticTypeInferenceVisitorTransformTest : VisitorTransformTestBase() {
             ),
             TestCase(
                 "custom function",
-                "format('test %d %s', 1, 'a')",
-                customFunctionSignatures = listOf(formatFunc.signature),
+//                "format('test %d %s', 1, 'a')",
```
Comment (johnedquinn) [inline partiql-lang/src/test/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransformTest.kt:7193]: Remove

File: partiql-lang/src/test/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransformTest.kt [hunk 17]
```diff
@@ -7623,8 +7623,8 @@ class StaticTypeInferenceVisitorTransformTest : VisitorTransformTestBase() {
 
             override val signature = FunctionSignature(
                 name = "format",
-                requiredParameters = listOf(StaticType.STRING),
-                variadicParameter = VarargFormalParameter(StaticType.ANY, 0),
+                requiredParameters = listOf(StaticType.STRING, StaticType.LIST),
+//                variadicParameter = VarargFormalParameter(StaticType.ANY, 0),
```
Comment (johnedquinn) [inline partiql-lang/src/test/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransformTest.kt:7627]: Remove

File: partiql-lang/src/test/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransformTest.kt [hunk 18]
```diff
@@ -7607,7 +7607,7 @@ class StaticTypeInferenceVisitorTransformTest : VisitorTransformTestBase() {
             val name: String,
             val originalSql: String,
             val globals: Map<String, StaticType> = mapOf(),
-            val customFunctionSignatures: List<FunctionSignature> = listOf(),
+            val customFunction: List<ExprFunction> = listOf(),
```
Comment (johnedquinn) [inline partiql-lang/src/test/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransformTest.kt:7610]:
```suggestion
            val customFunctions: List<ExprFunction> = listOf(),
```

File: partiql-lang/src/test/kotlin/org/partiql/lang/planner/transforms/LogicalToLogicalResolvedVisitorTransformTests.kt [hunk 19]
```diff
@@ -28,30 +28,31 @@ private fun PartiqlLogicalResolved.Builder.dynamicLookup(
...
-                )
-            ),
-            lit(
-                ionSymbol(
-                    when {
-                        globalsFirst -> "globals_then_locals"
-                        else -> "locals_then_globals"
-                    }
-                )
+//    searchTargets: PartiqlLogicalResolved.Expr,
```
Comment (johnedquinn) [inline partiql-lang/src/test/kotlin/org/partiql/lang/planner/transforms/LogicalToLogicalResolvedVisitorTransformTests.kt:31]: Remove

File: partiql-spi/src/main/kotlin/org/partiql/spi/functions/CustomInterface.kt
Comment (johnedquinn) [inline partiql-spi/src/main/kotlin/org/partiql/spi/functions/CustomInterface.kt:1]: Remove file

File: partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt [hunk 20]
```diff
@@ -169,6 +169,27 @@ public sealed class StaticType {
...
+            is BlobType -> superType is BlobType
+            is ClobType -> superType is ClobType
+            is StructType -> superType is StructType
+            is AnyOfType -> superType is AnyOfType
+            is DateType -> superType is DateType
+            is TimeType -> superType is TimeType
+            is GraphType -> superType is GraphType
+            else -> false
+        }
+    }
```
Comment (johnedquinn) [inline partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt:192]: Remove. See `org.partiql.lang.types.StaticTypeUtils.isSubTypeOf`

File: settings.gradle.kts
Comment (johnedquinn) [inline settings.gradle.kts:1]: Let's remove the entire `plugins/partiql-bananna`. Sorry if I didn't make that clear.

**Review by johnedquinn — CHANGES_REQUESTED** [review]
Good work. See other comments. However, overall we should:
1. share logic between variants of functions
2. update the CHANGELOG to include *all* public-facing changes.
3. remove the mock plugin
4. remove the query_power functions

I still need to look at the PhysicalPlanCompilerImpl, StaticTypeInferenceVisitorTransform, PlanTyper, EvaluatingCompiler, and FunctionManager.

File: partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt:1]: We'll need to modify the function map to be a function list in this file. And, we can now look into using the FunctionManager since we can pass in StaticTypes. I'll message offline about where we can place these tests.

**Review by johnedquinn — CHANGES_REQUESTED** [review]
This is a huge step in the right direction! Great work. On top of the changes requested, we should also add unit tests for the `FunctionManager`.

File: partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt [hunk 21]
```diff
@@ -548,15 +551,44 @@ internal object PlanTyper : PlanRewriter<PlanTyper.Context>() {
...
             return node.copy(type = StaticType.ANY)
         }
-        val type = node.type ?: computeReturnTypeForFunctionCall(signature, processedNode.args.getTypes(ctx), ctx)
+
+        var type = node.type
+        val funcsMatchingArity = functions.filter { it.signature.arity.contains(arguments.size) }
+        if (funcsMatchingArity.isEmpty()) {
+            handleIncorrectNumberOfArgumentsToFunctionCallError(funcName, getMinMaxArities(functions).first..getMinMaxArities(functions).second, arguments.size, ctx)
+        } else {
+            for (func in funcsMatchingArity) {
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt:567]: Similar to StaticTypeInferencer. Seems like we need to collect and union.

File: partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt:1]: With these changes, let's add trim overloading tests as part of the `PartiQLSchemaInferencerTests`. We can discuss offline.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Check argument types(ExprValue) by requiredArgs, OptionalArgs and VariadicArgs. This function will pass all arg values including MISSING or NULL values.
+     */
+    internal fun checkArgumentTypesEnableNullMissing(signature: FunctionSignature, args: List<StaticType>) {
+        fun checkArgumentType(formalStaticType: StaticType, actualStaticType: StaticType, position: Int) {
+            val formalExprValueTypeDomain = StaticTypeUtils.getTypeDomain(formalStaticType)
+
+            if (actualStaticType == StaticType.NULL || actualStaticType == StaticType.MISSING) {
+            } else if (!StaticTypeUtils.isSubTypeOf(actualStaticType, formalStaticType)) {
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt:51]:
```suggestion
            if (actualStaticType != StaticType.NULL && actualStaticType != StaticType.MISSING && !StaticTypeUtils.isSubTypeOf(actualStaticType, formalStaticType)) {
```

IMO, the lack of a statement in the IF body is confusing. Alternatively, you can just put a:
```kotlin
// Skip if NULL/MISSING
```

in the body of the IF.

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    /**
+     * Get minArity and maxArity by looping candidate functions filtered by function name.
+     */
+    internal fun getMinMaxArities(funcs: List<ExprFunction>): Pair<Int, Int> {
+        var minArity = Int.MAX_VALUE
+        var maxArity = Int.MIN_VALUE
+
+        funcs.forEach { func ->
```
Comment (johnedquinn) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/FunctionManager.kt:73]: See my other suggestion to make this a bit more compact. Not required, but could be nice.

File: CHANGELOG.md [hunk 24]
```diff
@@ -31,21 +31,30 @@ Thank you to all who have contributed!
 ### Added
 - Adds `org.partiql.value` (experimental) package for reading/writing PartiQL
   values
+- Adds function overloading to the `CompilerPipeline` and experimental `PartiQLCompilerPipeline`.
 
 ### Changed
+- Standardizes `org/partiql/cli/functions/QueryDDB` and other built-in functions by the new `ExprFunction` format.
+- **Breaking**: Redefines `org/partiql/lang/eval/ExprFunctionkt.call()` method by only invoking `callWithRequired` function.
+- **Breaking**: Redefines `org/partiql/lang/eval/builtins/DynamicLookupExprFunction` by merging `variadicParameter` into `requiredParameters` as a `StaticType.LIST` and renaming `callWithVariadic` function to `callWithRequired`.
```
Comment (johnedquinn) [inline CHANGELOG.md:39]:
```suggestion
- **Breaking**: Redefines `org/partiql/lang/eval/builtins/DynamicLookupExprFunction` by merging `variadicParameter` into `requiredParameters` as a `StaticType.LIST`. `callWithVariadic` is now replaced by`callWithRequired`.
```

File: CHANGELOG.md [hunk 25]
```diff
@@ -31,21 +31,30 @@ Thank you to all who have contributed!
...
+- **Breaking**: Modifies `functions` property of `CompilerPipeline`, `StaticTypeInferencer` and  `StepContext` to be a `List` and not a `Map`.
 
 ### Deprecated
 
 ### Fixed
 
 ### Removed
+- **Breaking**: Removes `optionalParameter` and `variadicParameter` from`org.partiql.lang.types.FunctionSignature`. To continue support for evaluation of `optionalParameters`, please create another same-named function. To continue support for evaluation of `variadicParameter`, please use a `StaticType.LIST` to hold all previously variadic parameters.
+  As this changes coincides with the addition of function overloading, only `callWithRequired` will be invoked upon execution of an `ExprFunction`.
+- **Breaking**: Removes unused class `Arguments` from `org.partiql.lang.eval.ExprFunction`.
```
Comment (johnedquinn) [inline CHANGELOG.md:49]:
```suggestion
- **Breaking**: Removes unused class `Arguments` from `org.partiql.lang.eval`.
```

File: CHANGELOG.md [hunk 26]
```diff
@@ -31,21 +31,30 @@ Thank you to all who have contributed!
...
+- **Breaking**: Removes unused parameter `args: Arguments` from `org.partiql.lang.eval.ExprFunctionkt.call()` method.
 
 ### Security
 
 ### Contributors
 Thank you to all who have contributed!
 - @howero
-- @<your-username>
+- @yuxiaotang
```
Comment (johnedquinn) [inline CHANGELOG.md:57]: Nit -- your contributing username is `@yuxtang-amazon`. This isn't blocking -- I'll leave it up to you which one you'd like to add to the CHANGELOG.

**Review by johnedquinn — CHANGES_REQUESTED** [review]
Overall, some really great work! The EvaluatingCompiler and other classes are *much* easier to comprehend as the logic for function resolution is abstracted away. On top of the comments left, let's also add some unit tests for the `FunctionManager` as well as integration tests for the `EvaluatingCompiler` and the `PartiQLSchemaInferencer`.

Thanks again! This is great work.

File: partiql-lang/src/test/kotlin/org/partiql/lang/eval/impl/FunctionManagerTest.kt [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+import org.partiql.lang.eval.builtins.SCALAR_BUILTINS_DEFAULT
+import org.partiql.types.StaticType
+
+class FunctionManagerTest : EvaluatorTestBase() {
+
+    val functions = SCALAR_BUILTINS_DEFAULT
+    private val functionManager = FunctionManager(functions)
+
+    @Test
+    fun functionManagerPassTests() {
```
Comment (johnedquinn) [inline partiql-lang/src/test/kotlin/org/partiql/lang/eval/impl/FunctionManagerTest.kt:18]: Great work on these tests.

1. Can you split each of these passing tests into separate tests?
2. Also, can you have test cases for passing in `null`?
3. Can you also have a test corresponding to a subtype? So, since `TRIM`'s first parameter is TEXT, can we attempt to resolve `TRIM(<string>)` ?

**Review by johnedquinn — CHANGES_REQUESTED** [review]

**Review by johnedquinn — APPROVED** [review]
Congrats on the great work!
